adserver/tasks.py

```diff
@@ -218,6 +232,128 @@ def daily_update_keywords(day=None):
                 )
 
 
+@app.task()
+def daily_update_regiontopic(day=None):  # pylint: disable=too-many-branches
+    """
+    Update the RegionTopicImpression index each day.
+
+    Each data point will have one region, but multiple possible topics.
+
+    :arg day: An optional datetime object representing a day
+    """
+    start_date, end_date = _get_day(day)
+
+    log.info("Updating RegionTopic's for %s-%s", start_date, end_date)
+
+    # Remove all old impressions, because they are cumulative
+    RegionTopicImpression.objects.filter(
+        date__gte=start_date,
+        date__lt=end_date,
+    ).delete()
+
+    for impression_type in ["views"]:
+        queryset = _default_filters(impression_type, start_date, end_date)
+
+        unsold = {}
+        no_keywords = {}
+
+        for values in (
+            queryset.values("publisher", "advertisement", "keywords", "country")
+            .annotate(total=Count("country"))
+            .filter(total__gt=0)
+            .order_by("-total")
+            .values(
+                "publisher",
+                "advertisement",
+                "keywords",
+                "country",
+                "advertisement__flight__targeting_parameters",
+                "total",
+            )
+        ):
+            if not (
+                values["keywords"]
+                and values["country"]
+                and values["advertisement__flight__targeting_parameters"]
+            ):
+                continue
+
+            keywords = json.loads(values["keywords"])
+            country = values["country"]
+            publisher_keywords = set(keywords)
+
+            flight_targeting = json.loads(
+                values["advertisement__flight__targeting_parameters"]
+            )
+            flight_keywords = set(flight_targeting.get("include_keywords", {}))
+
+            matched_keywords = publisher_keywords & flight_keywords
```

Comment:
Don't we want matched as well as unmatched keywords? Like just because a geo-targeted only ad is shown, we want to know if this could have been a data science ad.

Reply:
Do we? That will pretty quickly blow up the index if we're indexing every keyword on the publisher.

Reply:
Thinking more, I think you're right. We need this to get a real fill rate.

Reply:
What should we be doing with the unmatched keywords though? That is basically "keywords a publisher had, but didn't get an ad served against". It presumably shouldn't be part of the model record attached to the ad (since the ad wasn't served against that keyword) -- so should it have a blank advertisement?

Reply:
Basically what we want to know is how many `$TOPIC` ads we *could have* served in `$REGION`. If we got a matching ad request for `$TOPIC`/`$REGION` and chose to serve a more generic ad isn't relevant.